src/ast/pass_manager.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace bpftrace {
+
+class BPFtrace;
+
+namespace ast {
+
+class Node;
+class SemanticAnalyser;
+class Pass;
+
+/**
+   Result of a pass run
+ */
+class PassResult
+{
+public:
+  static PassResult Error(const std::string &msg);
+  static PassResult Success(Node *root = nullptr);
+
+  // Ok returns whether the pass was successful or not
+  bool Ok() const
+  {
+    return success_;
+  };
+
+  // Error returns the error generated by the pass, if any.
+  const std::string &Error()
+  {
+    return error_;
+  };
+  Node *Root() const
+  {
+    return root_;
+  };
+
+private:
+  bool success_;
+  std::string error_;
+  Node *root_ = nullptr;
+};
+
+/**
+   Context/config for passes
+
+   Note: Most state should end up in the BPFtrace class instead of here
+*/
+
+struct PassContext
+{
+public:
+  PassContext(BPFtrace &b) : b(b){};
+  BPFtrace &b;
+
+private:
+  // As semantic pass and map creation are separate passes we need this
+  SemanticAnalyser *semant;
+
+  // Only ones allowed to access semant
+  friend Pass CreateSemanticPass();
+  friend Pass CreateMapCreatePass();
+};
+
+using PassFPtr = std::function<PassResult(Node &, PassContext &)>;
+
+/*
+  Base pass
+*/
+class Pass
+{
+public:
+  Pass() = delete;
+  Pass(std::string name, PassFPtr fn) : fn_(fn), name(name){};
+
+  virtual ~Pass() = default;
+
+  PassResult Run(Node &root, PassContext &ctx)
+  {
+    return fn_(root, ctx);
+  };
+
+private:
+  PassFPtr fn_;
+
+public:
+  std::string name;
+};
+
+/**
+   Read only diagnostic AST passes
+ */
+class AnalysePass : public Pass
+{
+public:
+  AnalysePass(std::string name, PassFPtr fn) : Pass(name, fn){};
+};
+
+/**
+   AST modifying passes
+ */
+class MutatePass : public Pass
+{
+public:
+  MutatePass(std::string name, PassFPtr fn) : Pass(name, fn){};
+};
```

Comment:
Why does the pass manager need to be aware of _what_ each pass does? I'd expect the pass manager to just manage passes -- what each pass semantically does is up to pass. Furthermore, the caller shouldn't be using a pass unless they know what the pass does. Seems differentiating passes in the pass manager just adds friction.

Am I missing something?

Reply:
Yeah I don't think we really need it. I initially differentiated between them as mutating passes return a new root which needs handling, but ended up handling that by passing an optional arg to Success(). And because it would make it easy to fiilter  out modifying passes when debugging, but as passes are named we could just add a `skip pass` option.

The whole pass manager thing is experimental, not sure what interface will best suite our future needs. Just hoping we can keep it simple so refactoring in the future is easy.

Ill remove it :)